pkg/server/backend_manager.go

```diff
@@ -57,21 +57,38 @@ func newBackend(conn agent.AgentService_ConnectServer) *backend {
 	return &backend{conn: conn}
 }
 
-// BackendManager is an interface to manage backend connections, i.e.,
-// connection to the proxy agents.
-type BackendManager interface {
-	// Backend returns a single backend.
-	Backend() (Backend, error)
+// BackendStorage is an interface to manage the storage of the backend
+// connections, i.e., get, add and remove
+type BackendStorage interface {
 	// AddBackend adds a backend.
 	AddBackend(agentID string, conn agent.AgentService_ConnectServer) Backend
 	// RemoveBackend removes a backend.
 	RemoveBackend(agentID string, conn agent.AgentService_ConnectServer)
+	// NumBackends returns the number of backends.
+	NumBackends() int
+}
+
+// BackendManager is an interface to manage backend connections, i.e.,
+// connection to the proxy agents.
+type BackendManager interface {
+	// Backend returns a single backend.
+	Backend(ctx context.Context) (Backend, error)
```

Comment:
Can we add a warning that we should not be passing individual request context objects here? Finishing handling a request on the tunnel signifies that request is complete and *not* that the tunnel is complete.

Reply:
Yes. that is a very good point. I have added a warning comment.